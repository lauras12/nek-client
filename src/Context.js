import React from 'react';
import STORE from './store';


const HikeContext = React.createContext({

    currentHike: '',
    tracks: [],
    hikes: [],
    addHike: () => { },
    enterHike: () => { },
    updateFullHike: () => { },
    updateAttributes: () => { },
    tracksAttributes: [],
})

export default HikeContext;


export class HikeContextProvider extends React.Component {
    constructor() {
        super()
        this.state = {
            currentHike: null,
            tracks: STORE.tracks,
            hikes: STORE.hikes,
            trackAttributes: STORE.attributes,
            error: null,
        }
    }
    
    
    setError = (err) => {
        this.setState({
            error: err,
        })
    }
    
    setTracksList = (data) => {
        this.setState({
            tracks: data,
        })
    }

    
    enterHike = (hike) => {
        this.setState({
            currentHike: hike,
        })
    }

    addHike = (newHike) => {
        this.setState({
            hikes: [...this.state.hikes, newHike],
            currentHike: newHike,
        })
    }

    updateFullHike = (track, currentHike, sectionName) => {
        const hikeSection = currentHike[sectionName];
        const newHikeSection = [...hikeSection, track];
        const updatedHike = { ...currentHike, [sectionName]: newHikeSection };
        const updatedHikes = this.state.hikes.map(hike => {
            if(hike.id === currentHike.id) {
                return currentHike;
            } else {
                return hike;
            }
        })
        
        this.setState({
            currentHike: {...updatedHike},
            hikes: updatedHikes,
        });
    }

    updateAttributes = (newAttributes) => {
        this.setState({
            trackAttributes: [...this.state.trackAttributes, newAttributes],
        })
    }


    render() {
        const contextValue = {
            setError: this.setError,
            setTracksList: this.setTrackList,
            currentHike: this.state.currentHike,
            tracks: this.state.tracks,
            hikes: this.state.hikes,
            addHike: this.addHike,
            enterHike: this.enterHike,
            updateAttributes: this.updateAttributes,
            updateFullHike: this.updateFullHike,
            trackAttributes: this.state.trackAttributes,
        }

        return (
            <HikeContext.Provider value={contextValue}>
                {this.props.children}
            </HikeContext.Provider>
        )
    }
}