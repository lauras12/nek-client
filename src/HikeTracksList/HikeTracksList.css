.hike-tracks-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around; 
    margin: auto;
    padding-bottom: 20px;
}

@media (min-width: 600px) {
    .hike-tracks-container {
        display: flex;
        flex-wrap: wrap;
        justify-content: space-around; 
        width: 80%;
        margin: auto;
        padding-bottom: 20px;
    }
}  
