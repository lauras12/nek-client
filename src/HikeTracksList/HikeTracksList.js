import React from 'react';
import TrailContext from '../Context';
import HikeItem from '../HikeItem/HikeItem';
import APIHikeCalls from '../services/API_Hike_service';
import './HikeTracksList.css';


export default class HikeTracksList extends React.Component {
    static contextType = TrailContext;
    state = {
        error: null,
    }

    componentDidMount = () => {
        let hikeId = this.context.currentHikeId;
      
        if(!hikeId) {
            this.props.history.push('/')
        } else {
            APIHikeCalls.getAllTrackInHike(hikeId)
            .then(data => {
                this.context.setCurrentHike(data);
            })
            .catch(err => {
                this.setState({
                    error: err
                });
            });
        }
        
    }

    render() {
        let currentHikeTracksIds = this.context.currentHike.assignedTracks;
        const { tracks } = this.context;

        let orderedIds = currentHikeTracksIds.map(element => {
            return element.map(id => {
                return tracks.find(track => track.id === id);
            });
        });
           
        
        const hikeTracks = orderedIds.map(element => element.map((track, index) => {
            return (
                <HikeItem
                    key={index}
                    {...this.props} 
                    trackId={track.id}
                    img={track.img}
                    hikeId={this.context.currentHike.id}
                />
            );
        }));
        
        return (
            <div >
                 <div className='error'>
                    {this.state.error ? this.state.error.message : null}
                </div>
                <ul className='hike-tracks-container'>
                {hikeTracks}
                </ul>
            </div >
        );
    }
}