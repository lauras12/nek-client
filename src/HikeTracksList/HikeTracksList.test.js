import React from 'react';
import ReactDom from 'react-dom';
import renderer from 'react-test-renderer';
import HikeTracksList from './HikeTracksList';
import {BrowserRouter as BR} from 'react-router-dom';
import { HikeContextProvider as TCP } from '../Context';

describe('HikeTracksList component', () => {
    const props = {
        match: {params: {path: '/hike'}},
        history: {push: () => {}}
    };
    
    it ('renders without crashing',() => {
        const div = document.createElement('div');
        ReactDom.render(<BR><TCP><HikeTracksList {...props} /></TCP></BR>, div);
        ReactDom.unmountComponentAtNode(div);
    });

    it('renders UI as expected', () => {
        const loginPage = renderer.create(<BR><TCP><HikeTracksList {...props}/></TCP></BR>);
        expect(loginPage.toJSON()).toMatchSnapshot();
    });
});