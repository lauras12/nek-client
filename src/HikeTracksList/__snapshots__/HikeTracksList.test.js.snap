// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HikeTracksList component renders UI as expected 1`] = `
<div>
  <div
    className="error"
  />
  <ul
    className="hike-tracks-container"
  />
</div>
`;
